soroban-env-host/src/host/metered_clone.rs

```diff
@@ -121,82 +185,256 @@ impl MeteredClone for ScVal {
+    const ELT_SIZE: u64 = C::ELT_SIZE + 8;
+
+    fn charge_for_clone(&self, budget: &Budget) -> Result<(), HostError> {
+        // first take care of the Option part, which is just a shallow copy
+        charge_shallow_copy::<Self>(1, budget)?;
+        // then take care of any substructure clone (recursively)
         match self {
-            Some(elt) => Ok(Some(elt.metered_clone(budget)?)),
-            None => Ok(None),
+            Some(elt) => elt.charge_for_clone(budget),
+            None => Ok(()),
+        }
+    }
+}
+
+impl MeteredClone for DebugEvent {
+    const IS_SHALLOW: bool = false;
+
+    const ELT_SIZE: u64 = 80;
+
+    fn charge_for_clone(&self, budget: &Budget) -> Result<(), HostError> {
+        charge_shallow_copy::<Self>(1, budget)?;
+        if self.args.is_heap() {
+            // equivalent to charging for a `Vec` clone.
+            charge_shallow_copy::<Vec<DebugArg>>(1, budget)?;
+            charge_heap_alloc::<Vec<DebugArg>>(self.args.len() as u64, budget)?;
+            charge_shallow_copy::<Vec<DebugArg>>(self.args.len() as u64, budget)
+        } else {
+            // equivalent to charging for a slice clone.
+            DebugArg::charge_for_clones(self.args.as_slice(), budget)
+        }
+    }
+}
+
+impl MeteredClone for ContractEvent {
+    const IS_SHALLOW: bool = false;
+
+    const ELT_SIZE: u64 = 104;
+
+    fn charge_for_clone(&self, budget: &Budget) -> Result<(), HostError> {
+        charge_shallow_copy::<Self>(1, budget)?;
+        let ContractEventBody::V0(event) = &self.body;
+        event.topics.charge_for_clone(budget)?;
+        event.data.charge_for_clone(budget)
+    }
+}
+
+impl MeteredClone for HostEvent {
+    const IS_SHALLOW: bool = false;
+
+    const ELT_SIZE: u64 = 112;
+
+    fn charge_for_clone(&self, budget: &Budget) -> Result<(), HostError> {
+        match self {
+            HostEvent::Contract(c) => c.charge_for_clone(budget),
+            HostEvent::Debug(d) => d.charge_for_clone(budget),
         }
     }
 }
 
 impl MeteredClone for Events {
     const IS_SHALLOW: bool = false;
-    fn metered_clone(&self, budget: &Budget) -> Result<Self, HostError> {
-        budget.charge(CostType::CloneEvents, self.0.len() as u64)?;
-        Ok(self.clone())
+
+    const ELT_SIZE: u64 = 24;
+
+    fn charge_for_clone(&self, budget: &Budget) -> Result<(), HostError> {
+        self.0.charge_for_clone(budget)
+    }
+}
+
+impl MeteredClone for InternalEvent {
+    const IS_SHALLOW: bool = false;
+
+    const ELT_SIZE: u64 = 80;
+
+    fn charge_for_clone(&self, budget: &Budget) -> Result<(), HostError> {
+        charge_shallow_copy::<Self>(1, budget)?;
+        match self {
+            InternalEvent::Contract(c) => c.charge_for_clone(budget),
+            InternalEvent::Debug(d) => d.charge_for_clone(budget),
+            InternalEvent::None => Ok(()),
+        }
+    }
+}
+
+mod test {
+    #[allow(unused)]
+    use super::*;
+
+    #[test]
+    fn test_elt_size() {
```

Comment:
I'm not sure if there's a way to automatically cover this test on all types implementing the `MeteredClone` trait.